pkg/controller/grouppermission/grouppermission_controller_test.go

```diff
@@ -0,0 +1,380 @@
+// given: clusterRoleName, groupName
+// want: a ClusterRoleBinding that contains the new clusterRoleName and groupName
+func TestClusterRoleBinding(t *testing.T) {
+	ctx := context.TODO()
+	reconciler := newTestReconciler()
+	defer reset(ctx, reconciler)
+
+	// Register operator types with the runtime scheme.
+	s := scheme.Scheme
+
+	//Add api to scheme
+	if err := operatorapi.AddToScheme(s); err != nil {
+		t.Fatalf("Unable to add operatorapi scheme: (%v)", err)
+	}
+
+	// creates a groupPermission object
+	nerr := reconciler.client.Create(ctx, makeGroupPermission())
+	if nerr != nil {
+		t.Errorf("Couldn't create required GroupPermission object for test: %s", nerr)
+	}
+
+	// create a request - mock the API call
+	request := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      "test-group-permission",
+			Namespace: "rbac-permissions-operator",
+		},
+	}
+
+	// this is the function we are testing
+	// it should return mockClusterRoleBinding() which contains the same clusterRoleName and GroupName
+	newClusterRoleBinding := newClusterRoleBinding("exampleClusterRoleName", "exampleGroupName")
+
+	// compare the two clusterRoleBinding. They should be exactly the same
+	// if not our test fails, log out the difference
+	diff := reflect.DeepEqual(*newClusterRoleBinding, *mockClusterRoleBinding())
+	if !diff {
+		t.Error(diff)
+	}
+
+	// process the request
+	res, err := reconciler.Reconcile(request)
+	if res.Requeue {
+		t.Error("Didn't expect to requeue")
+	}
+	if err != nil {
+		t.Errorf("Got an unexpected error: %s", err)
+	}
+}
+
+// TestBuildClusterRoleBindingCrList tests buildClusterRoleBindingCrList function
+// given: GroupPermission Spec
+// want: slice of ClusterRoleBindingNames which consist of clusterrolename-groupname
+func TestBuildClusterRoleBindingCrList(t *testing.T) {
+
+	// this is the function we are testing by using a mock
+	buildList := buildClusterRoleBindingCRList(mockGroupPermissionSpec())
+
+	// this is the result that we want
+	result := []string{"exampleClusterRoleName-exampleGroupName", "exampleClusterRoleNameTwo-exampleGroupName"}
+
+	// check to see if desired is equal to result
+	if len(buildList) != len(result) {
+		t.Errorf("the length does not match")
+	}
+	for i, v := range result {
+		if v != buildList[i] {
+			t.Errorf("got %s, want %s", buildList, result)
+		}
+	}
+}
+
+// TestUpdateCondition tests the updatecondition function.
+// given: GroupPermission object, message, clusterRoleName, status, and state
+// want: an updated GroupPermission object with the correct updated fields
+func TestUpdateCondition(t *testing.T) {
+	time := metav1.Now()
+
+	// this is the function we are testing with a mock
+	buildCondition := updateCondition(mockGroupPermissionStatus(time), "testMessage", "testClusterRoleName", false, "testState")
+
+	// make a map of the result that we want to check mock against
+	testMap := make(map[int]managedv1alpha1.Condition)
```

Comment:
The more traditional way to define this structure is something like:

```go
var tests = []struct{
  teststruct managedv1alpha1.Condition, // test input
  challenge managedv1alpha1.Condition,  // what to compare it against
  testResult bool,                           // do we expect this test to return true or false?
}{
  { structInstance1, challengeStruct1, true},
  { structInstance2, challengeStruct2, false},
}
```

This way it's a bit easier to add more tests. Iterate over:

```go
for _, test := range tests {
  if theFunctionBeingTested(test.teststruct,test.challenge) != test.testResult {
    t.errorF("test failed\n")
  }
}